plugin/videofilterrunnable.cpp

```diff
@@ -0,0 +1,112 @@
+/**
+ * @file videofilterrunnable.cpp
+ * @copyright 2020 Open Mobile PLatform Ltd.
+ * @author Dmitry Butakov d.butakov@omprussia.ru
+ */
+
+#include "videofilterrunnable.h"
+
+#include <QDebug>
+#include <QtDBus/QDBusConnection>
+
+#include <sys/mman.h>
+#include <sys/stat.h>
+#include <fcntl.h>
+#include <unistd.h>
+
+VideoFilterRunnable::VideoFilterRunnable(QrFilter *filter) : m_filter(filter)
+{
+    connect(this, &VideoFilterRunnable::stringFound,
+            m_filter, &QrFilter::setResult, Qt::QueuedConnection);
+    // analyze empty frame to establish D-Bus connection
+    analyze(QDBusUnixFileDescriptor(0), 0,
+            QVideoSurfaceFormat(QSize(0,0), QVideoFrame::PixelFormat::Format_BGRA32));
+}
+
+static QString getRandomString()
+{
+   const QString possibleCharacters("ABCDEFGHIJKLMNOPQRSTUVWXYZ"
+                                    "abcdefghijklmnopqrstuvwxyz0123456789");
+   const int randomStringLength = 12;
+
+   QString randomString;
+   for (int i = 0; i < randomStringLength; ++i) {
+       int index = qrand() % possibleCharacters.length();
+       QChar nextChar = possibleCharacters.at(index);
+       randomString.append(nextChar);
+   }
+   return randomString;
+}
+
+QVideoFrame VideoFilterRunnable::run(QVideoFrame *input, const QVideoSurfaceFormat &surfaceFormat,
+                                     QVideoFilterRunnable::RunFlags flags)
+{
+    Q_UNUSED(flags)
+
+    auto in = !input ? QVideoFrame() : *input;
+    if (!m_reply.isFinished()) {
+        return in;
+    }
+
+    if (m_reply.isError()) {
+        qWarning() << "QR Filter error" << m_reply.error().message();
+    } else {
+        QString msg = m_reply.value();
+        if (!msg.isEmpty()) {
+            emit stringFound(msg);
+        }
+    }
+    m_reply = QDBusPendingReply<QString>();
+
+    if (!in.isValid()) {
+        return in;
+    }
+
+    if (!in.map(QAbstractVideoBuffer::ReadOnly)) {
+        return in;
+    }
+
+    auto pBits = reinterpret_cast<char*>(in.bits());
+    auto bufferSize = static_cast<uint>(in.mappedBytes());
+
+    // Generate tmp frame buffer and unlink immediately to avoid frame collision
+    const char *tmpFrameBufferName =  getRandomString().toStdString().c_str();
+    int fd = shm_open(tmpFrameBufferName, O_RDWR | O_CREAT, 0777);
+
+    if (fd != -1) {
+        shm_unlink(tmpFrameBufferName);
+        ssize_t wr = write(fd, pBits, bufferSize);
+        if (wr > 0) {
+            if (static_cast<size_t>(wr) == bufferSize) {
+                QDBusUnixFileDescriptor dbus_fd(fd);
+                analyze(dbus_fd, bufferSize, surfaceFormat);
```

Comment:
Somehow had the impression that we had added some maximum frequency for the analysis but indeed only now limited by pending calls. For testing added a QElapsedTimer and "qDebug() << m_elapsedTimer.restart()" to check the frequency. On my test device it's commonly under 100ms before the next request is made.

With CPU usage mattering, should we still have some maximum frequency? Should be easy to do with the elapsed timer. Wouldn't consider unreasonable to require targetting a qr code for a second or two even, but some 1-5 analysis per second could suffice?

Reply:
Actually on R7 I have same numbers, added timer which reduces frame analysis to 2 per second.

Reply:
Thanks! Greatly improved the cpu usage and the detection still feels semi instant.